Refactor fonts

McGiverGim opened the pull request:
Reorder some symbols, delete symbols not used, added some new...

The user @Jetrell has made this great work. I copy&paste his thoughts:

----------------------------------
This is the cleanup of all the fonts. With some relocations.
It has 8 arrow characters removed. With 4 generic arrows added, across all the fonts.
The `AHI center` characters have all been relocated together.
And an `Over Home` symbol has been added.

1. 0x05 `Up Triangle` is removed. Replaced with `Over Home` symbol.
2. 0x7b `AH_CENTER_LINE` is relocated to 0x72 - leaving 0x7b blank
3. 0x7e `AH_CENTER` is relocated to 0x73 - leaving 0x7e blank
4. 0x7d `AH_CENTER_LINE_RIGHT` is relocated to 0x74 - leaving 0x7d blank
5. 0x79 previous `arrow set`, has also been removed and left blank
6. 0x75 is an `UP` arrow
7. 0x76 is a `DOWN` arrow
8. 0x77 is a `RIGHT` arrow
9. 0x78 is a `LEFT` arrow

![image](https://user-images.githubusercontent.com/2673520/58938783-f56d1180-8775-11e9-9f9a-83e55561d6bd.png)

There are 5 variants of the `Over Home` symbol that are used across the fonts, to give them individuality.
Here is a sample of 3 of them.
Hope they are okay for use in #8372 @McGiverGim
![image](https://user-images.githubusercontent.com/2673520/58938863-1d5c7500-8776-11e9-9d4c-18a36152519d.png)
![image](https://user-images.githubusercontent.com/2673520/58938906-336a3580-8776-11e9-9cd2-5a6af2bc41a7.png)
![image](https://user-images.githubusercontent.com/2673520/58938912-36652600-8776-11e9-896d-6261c9b9a209.png)

The 4 generic arrows that have been added, are laid out in a way, to allow them to be used individually or collectively in future developments.
e.g. 3D throttle graph/scale or even an accelerometer graph, for showing directional motion etc.
![image](https://user-images.githubusercontent.com/2673520/58938937-43821500-8776-11e9-836a-c0952cb3b6b6.png)

If I have missed anything... let me know:-)

mikeller commented:
I like the cleanup, but I am not sure about the 'home' symbol that is a 'H' in a circle - to me this is a 'hospital' symbol. How about a stylised house (no line above it needed), or a circle with a stylised house in it?

Jetrell commented:
@mikeller  Not all have the `H `symbols in a cycle. Some have a `H` in an down arrow. @McGiverGim like the idea of that one. 
Do you have any preference in which to place the House in a circle?

It made me laugh... when you said a stylized house. There won't be enough room in this tiny house to turn around :-)
If you like it.... I'll re upload the fonts again tomorrow, in the place of today's, in PR 1392.
![home symbol](https://user-images.githubusercontent.com/47995726/58965577-268c2700-87f4-11e9-8606-b8ccc61195fd.PNG)

mikeller commented:
@Jetrell: Looks good!

mikeller commented:
@Docteh: Yes, you are right. Somehow I got that mixed up with the square with a 'H' in it: https://www.shutterstock.com/search/hospital+road+sign

As you say, they often come together. :wink:

Jetrell commented:
I ended up placing this home symbol in all the fonts. @mikeller It looked far better in the goggles, than the one I previewed yesterday.
![Home symbol](https://user-images.githubusercontent.com/47995726/59000939-1a35b780-8850-11e9-9f20-981c0214d29d.PNG)

[Font cleanup.zip](https://github.com/betaflight/betaflight-configurator/files/3259630/Font.cleanup.zip)

McGiverGim commented:
Updated the fonts with the latest version.

mikeller commented:
@McGiverGim, others: Is this good for a merge now? :-)

McGiverGim commented:
To me is ok, but it will need changes in firmware and preview. Maybe I can do it tomorrow if we want to merge all of them at the same time.

McGiverGim commented:
@mikeller added a commit to make the preview compatible with the new order of symbols and with different versions, but I don't think this is the correct approach.

Each element preview is only adapted to the latest version, for example the speed element is in the form of `SXXXK`. This is ok when using the new fonts in the preview, but if we use the older the S initial contains a different symbol than speed and looks very strange. 

If the idea is that the OSD preview is a best effort, at least the element must be "correct" in the preview, maybe not the same than in the goggles, but at least not show strange symbols.

If we want to maintain only a preview for each element, what about:
- Upgrade the fonts in the FC depending on the version, like until now.
- For the preview, use always the latest font version available, in this way the previews will be more or less correct (depending on the version) and not have strange symbols.

McGiverGim commented:
Playing a little with it, it will produce problems with the custom fonts uploaded by the user :( They will not appear in the preview.

Another solution is to fix the previews with problems, pointing to a empty char if it does not exist in version 1 or maintaining two previews different depending on the version of Betaflight.

McGiverGim commented:
Finally I found a workaround... extended the preview function to let use an empty var `''` or `null` as value to define an empty element. In this way only we need to maintain the definition of symbols by version and only for those symbols that does not exist in previous versions.
We must use the preview in form of a function and not of a string too, because:
- The previews are loaded when loading the Configurator and the CONFIG var does not exist at this moment.
- If changed the FC the previews will not reload if they are not a function.

Docteh commented:
It makes sense that lots/all previews should be functions. How come some of the SYM items are? I figured at some point the whole section of setting SYM would be moved to the tab load.

simonbethke commented:
That's an H in a circle. Its the german sign for a bus stop:
https://www.schilder-versand.com/assets/gfx/product/500x500/haltestelle-verkehrsschild-nr-224-50-2898.jpg

mikeller commented:
@simonbethke: Hehe, good one.

@McGiverGim: I am not totally convinced that it makes sense to start to support different versions of fonts for previews - we have not done this in the past when we changed how elements are displayed, and I don't think we ever had complaints about it:
- the preview is meant to be just that - a preview to allow the user to arrange the elements and check that the arrangement looks ok. It is not meant to be a 'reference' for what will appear in the OSD;
- the only value I can see in supporting versioned element formatting in the preview is in cases where the length / size of an element is changed - in this case users might be tripped up by neatly aligning the preview, and then getting overlap. But since a lot of the elements do not have a fixed length anyway, the current preview is not a guarantee that this won't happen anyway;
- OSD preview elements is one thing where we are often struggling with keeping the configurator up to date as it is right now. I think that adding more complexity, and more requirements to keep things synchronised, will only make this problem worse.

Thoughts?

McGiverGim commented:
@mikeller I didn't change the preview exactly. I left the preview "format" but I changed the symbols when necessary because the result in the preview is totally wrong (not bad format, bad appearance) for older versions:
- For example, here you have the preview for the crosshair (some arrows) and the speed (on hr) when you use Betaflight 4.0 (for example all F3 FC) with the new Configurator:
![image](https://user-images.githubusercontent.com/2673520/59156608-28791300-8a9e-11e9-9292-a8ecf4ab47a0.png)
- What I have observed is that maybe I have changed too much symbols. This is only needed for symbols moved or symbols that does not exist in the v1 version. Some of them were added to the latest version of the v1 fonts, so I can remove them and simplify a little the maintenance.
- What I've done in this PR, is the same that it was done in the font versions PR here: 
https://github.com/betaflight/betaflight-configurator/blob/5694ac62a888ef44c97eca90e5a2efeaf47efed9/src/js/tabs/osd.js#L2459-L2466 
but I have simplified it doing the customization in the same definition, in this way nobody needs to search into the code for the place where to change it if needed.

So to me is clear that some symbols need to take into account the version, but maybe I changed too much of them and I can remove the customization where not needed, limiting it to the minimum necessary. For new OSD elements it will not be necessary because they will not appear for older versions of firmware, so this will not require too much maintenance (if we no decide to use a v3 font, of course). What do you think?

mikeller commented:
@McGiverGim: I see. I am not sure if this is the right way to do it then. We could probably get away with an easier implementation if we just always used the latest font version, and made the preview work for the latest version.
Additionally we could add a warning somewhere on the tab that is shown if a user is using an older firmware version that tells them something like 'Preview shown for the latest version of Betaflight. Some elements might look different than the preview in the version that you are using.'

McGiverGim commented:
The problems when using the latest version always, were two:
- The implementation is not very elegant. The way the OSD is made don't let it in a clean way and easy. We have previews in the font manager that use the same font than the OSD preview.
- If the user uploads it's own font, the preview will not work or will not show what it must.

I prefer to change the char code for only two OSD elements by version and let all functional. In the actual code we have the same but not as clean as this implementation.

Docteh commented:
You could regenerate the v1 fonts if that makes things easier. The characters being replaced so far have not been used, which is why its mostly okay to stick them onto the old fonts.

These are the exceptions/instructions for making a v1 font.
![0x05](https://user-images.githubusercontent.com/190571/59160650-0aef7d80-8a8d-11e9-8cf3-d0829582fc23.png) take a triangle and copy to 0x05
0x72 AH_CENTER_LINE copy to 0x26
0x74 AH_CENTER_LINE_RIGHT copy to 0x27

Jetrell commented:
> 
> I prefer to change the char code for only two OSD elements by version and let all functional. In the >actual code we have the same but not as clean as this implementation.

@McGiverGim 
I suppose that now is time this will matter the most. Between the 4.0 and 4.1 release?
I figure that most people that are using nightly builds. Will keep up with changes... So the V2 fonts from  a couple of weeks ago, will soon be replaced with the V3 and its previews?

Jetrell commented:
> 
> I prefer to change the char code for only two OSD elements by version and let all functional. In the >actual code we have the same but not as clean as this implementation.

@McGiverGim 
I suppose that now is the time this will matter the most. Between the 4.0 and 4.1 release?
I figure that most people that are using nightly builds. Will keep up with changes... So the V2 fonts from  a couple of weeks ago, will soon be replaced with the V3 and its previews?

Docteh commented:
I figure any changes between now and the code freeze for 4.1 counts as V2.

I should ask about when the code freeze is.

McGiverGim commented:
Exactly, between the 4.0 and 4.1 we change from v1 fonts to v2 fonts. We don't have (and there are no plans for it at this moment) a v3 font. The latest change in fonts simply have overwritten the v2 font because it was not released until this moment. The nightly firmwares must always use the latest nightly Configurator with the latest fonts.

The font version only needs to change when some used element changes its position in the font. Adding new elements at unused/blank positions did not produce a new font version. After this refactor we have freed some positions so I don't expect a v3 near in the future.

The major problem with 4.0 is that it will be the latest version for F3 CPU, so it will be used for a long time. For this reason I want to give something usable to them. I don't need to show in the preview the "real" element, but at least a distinguishable one.

McGiverGim commented:
Modified the PR with the minimum symbols needed to be changed. Only crosshair and speed prefix.

The crosshair was being changed in another place of the code, so the only real change is the speed prefix one, and a little refactor of the code to do it in the definition of the symbols to be easier to find.

I'm playing a little too with a branch in local to see if I can find an easy (and not too ugly) way to use always the latest font, but this will break the preview of user uploaded fonts for Betaflight 4.0 or earlier.

McGiverGim commented:
Sorry, I forgot to clean some unneeded functions preview. New PR pushed. Now the code, as you can see, is very simple.

mikeller commented on src/js/tabs/osd.js:
```diff
@@ -23,7 +22,7 @@ SYM.METRE = 0xC;
 SYM.KPH = 0x9E;
 SYM.MPH = 0x9D;
-SYM.SPEED = 0x70;
+SYM.SPEED = function() { return semver.gte(CONFIG.apiVersion, "1.42.0") ? 0x70 : '' };
```
I'd just return `SYM.BLANK` in the second case here, and avoid the check for empty string below.

McGiverGim replied:
I did it in this way to try to let a more accurate preview. We need to customize the symbol in a version when this symbol does not exist in the older font version. Usually this implies that the OSD in the firmware does not print nothing. Putting a blank can let a strange preview if the symbol is in the middle, for example the ESC temperature, and will add spaces to the left that will make the preview very different than what the goggles show in others situations like the speed. I know the user must verify it before flying but I think this a better "best effort" for the OSD at very little cost.

mikeller replied:
Not sure I understand that right. If we are trying to massage the characters we use so that a message that we are printing in the format used for one version looks like the message for another version then this will be very hard to maintain, and should be avoided.
If we want to support historical versions of a message, and the message format (as opposed to the character set used to print that message) was changed, then we need to store the historical formats in addition to the latest format.

McGiverGim replied:
No, we are not trying to maintain historical message formats. I'm trying to display always the latest message format (as we have always done) but replacing "incorrect/ugly" characters by the most correct one. If the character exist (but was not used for this version) I let it, the user can identify the string and must verify in the goggle how it looks in its version, but if the character does not exist or it has been replaced by an incorrect character I try to fix it or remove it.

Docteh replied:
I think not printing certain characters is fine. They're being dropped after the message is formatted, so it's not a huge thing. And with the disclaimer on previews. Half-assing is okay

mikeller replied:
In this case I'd probably return a blank for the character we cannot print, or maybe even a visible placeholder for it, like a `*`. Completely removing the character will just change the alignment with other lines, and make it look even messier.

McGiverGim commented:
This was not done in a block, like in other places to not modify a lot the file. In the OSD, the code for each symbol is loaded when importing the file in the main html. And the previews for all elements are created too then (except if they are treated as a function) so without refreshing all the previews the change of the symbols will do nothing for all the elements that haven't it defined as a function.
The PR that created the versioning of the fonts did something similar to what you say:
https://github.com/betaflight/betaflight-configurator/blob/5694ac62a888ef44c97eca90e5a2efeaf47efed9/src/js/tabs/osd.js#L2459-L2466
The problem of this code, like you can see, is that we define the symbols twice, one at the beginning and another one here, and the value too, because we need to reset it here to the original value too when the version is old.
So we will need to move the entire block of defined elements to this part of code, to not have it twice and it will be a more difficult to find them and less cleaner. I can put all of them inside a function too here at the beginning and call it from the loading of the code too, and maybe try to refresh all the previews, but I think that the code of the OSD is complicated enough to increment it more, but is a possibility.
So I think we have three solutions without doing a complete refactor of the OSD code:
- The solution of this PR. Is not perfect, but all works and we don't break anything.
- Include the modified symbols in a block. Similar to the first one but we will need to be sure than the symbols and previews are refreshed each time the tab is loaded.
- Use always the latest font for the preview. The code will be a hack to make this, and we break the possibility to preview a custom font for the user.

I'm open to any of the changes, but all of them have some cons. I only want to let some solution that does not break any feature that now is working.

mikeller commented:
@McGiverGim:

> The solution of this PR. Is not perfect, but all works and we don't break anything.

It works and is maintainable _for now_. Once we have half a dozen versions we'll be having a lot of conditional functions throughout all of the symbol definitions, and it will be very hard to debug and figure out what symbols are used for any given version.

> Use always the latest font for the preview. The code will be a hack to make this, and we break the possibility to preview a custom font for the user.

I don't see how this would be a hack. Can you explain?

Also, I am not sure if it is worthwhile to go out of our way to enable users to preview custom fonts with deprecated versions of the firmware - I think it is a good approach to make configurator _not break_ with deprecated versions, but that does not mean that we have to support these versions to the same level that we support the latest version - after all, we do want to motivate the users to move to the latest version.

McGiverGim commented:
> It works and is maintainable for now. Once we have half a dozen versions we'll be having a lot of conditional functions throughout all of the symbol definitions, and it will be very hard to debug and figure out what symbols are used for any given version.

We have needed several years to get from v1 to v2. I don't know if we must fix it now or go for the simpler solution. If you want that I fix it now I can do it like a said: refreshing the symbols and the previews each time the tab is loaded.

> I don't see how this would be a hack. Can you explain?

The solution is not the hack. It is a good solution (I tried it before at this PR). The hack is how we need to do it in the code if we don't change a lot of things. Now, each time the user changes the font in the select input, the code loads the new font, depending of the version, and executes a refresh of the custom logo, the font manager preview and the OSD preview. To make this solution work, without too much change, for a v1 font firmware we need to: load v1 font, refresh custom logo and font manager preview, load a v2 font, refresh OSD preview and let the v2 font loaded (the user can add more elements to the OSD preview and we need to refresh it). If the user uploads the font, we will need to reload the v1 font, upload it, and reload the v2 font to let it loaded for the OSD preview. I have tested only part of this, so I don't know if all will work. Maybe there is an easy way with only few changes (maybe load two versions of the font in the object and try to use them accordingly) but actually the immediate solution is like this.

mikeller commented:
> We have needed several years to get from v1 to v2. I don't know if we must fix it now or go for the simpler solution. If you want that I fix it now I can do it like a said: refreshing the symbols and the previews each time the tab is loaded.

We can go for this now, so that this can be merged, and fix it up later.
But I stand by my statement that the solution to selectively replace global 'constants' with functions is a maintainable approach - since the mapping isn't constant any more we should re-do it in a more appropriate way.

> Now, each time the user changes the font in the select input, the code loads the new font, depending of the version, and executes a refresh of the custom logo, the font manager preview and the OSD preview.

Ok, I had not considered that. This will add a lot more complexity, making it not attractive at all.

McGiverGim commented:
> We can go for this now, so that this can be merged, and fix it up later.

I will give a try to the refreshing think this week. If it works I will push another PR.

McGiverGim commented:
@mikeller finally, I was half wrong. There is no need to reload the preview, only the symbols. So I added here the change to reload the symbols when the tab is loaded. I have tested it and seems to work right.

Jetrell commented:
@McGiverGim I have spent some time over the last few days flying/watching DVR footage, trying to perfect these 3D home arrows in the Impact fonts. After many flights and much DVR watching and changes. There are only two that I could make a little better. (0x66 and 0x6a) I could picture them here... But the change doesn't show as well in a PNG as it does in the goggles. 
Could you please push them with these mods. 
Thanks
[Impact fonts 3D arrow improvements.zip](https://github.com/betaflight/betaflight-configurator/files/3275823/Impact.fonts.3D.arrow.improvements.zip)

McGiverGim commented:
@Jetrell I suppose these are only for V2 version or you will modify the V1 version too?

Jetrell commented:
> 
> 
> @Jetrell I suppose these are only for V2 version or you will modify the V1 version too?

Yep. They are for the V2. If you want them for the V1... I can do that too. 
 i.e. V1 meaning, the ones before this refactoring?  I have some many groups from over the last couple of months. And I have them label for my reference. So i just want to be sure which one you mean :-)

McGiverGim commented:
We have two versions: V1 for Betaflight 4.0 and earlier, and V2 for Betaflight 4.1 and later. 

You can grab the actual V1 here: https://github.com/betaflight/betaflight-configurator/tree/master/resources/osd

The V1 is like the V2 before all the refactoring and moving of crosshairs.

Jetrell commented:
@McGiverGim Here they are. Check it over and see if I missed anything. Because Impact fonts wern't released in this configurator.
The main thing I changed differently was 0x21. The 100 symbol. It was never used, but having "!" will fix the !FS! warning with 4.0 and earlier for anyone that uses Impact fonts with V1.

[Impact fonts V1.zip](https://github.com/betaflight/betaflight-configurator/files/3276474/Impact.fonts.V1.zip)

![Capture](https://user-images.githubusercontent.com/47995726/59272637-c684fd00-8c99-11e9-8b65-593917873e8f.PNG)

McGiverGim commented:
Pushed upgraded fonts.
